### PR TITLE
Fixes #1082 - add to cart from listing if only one attribute

### LINKED ISSUE #1082: Ability to add items with attributes from listing page leads to attributes not set
Adding items from Downloads category 

index.php?main_page=index&cPath=54_60

can be done from the listing page, but it leads to items in the cart without attributes correctly set.

### PR DESCRIPTION
**Pull request by scottcwilson**
Right now any number of attributes are permitted, which means the add is not correctly done; attributes are not set in cart.
Also included - new product in demo data that shows this capability.

### CODE REVIEW

**Comment by drbyte** [comment]
👍  to the query change

Not really sure how the table-truncations made it into this PR. That seems very out of scope.

**Comment by scottcwilson** [comment]
I can separate those out if you prefer; I added them to make it easier to test the demo products.

**Comment by drbyte** [comment]
What are the downsides of doing this truncation? Potential risks? Impact on other devs who've been used to it not truncating for years.  This discussion is totally out of scope, but I'd ask the same question if you split it into another PR.

**Comment by scottcwilson** [comment]
Let me split it to a second PR and make it *separate* from the  zc_install/sql/demo/mysql_demo.sql script.  Then only folks who desire this cleanup capability will use it.  It's just nice to have to make it easier to build out the demo products and do repeated testing while tweaking.

**Comment by scottcwilson** [comment]
LOL stupid close button!

**Comment by drbyte** [comment]
👍

File: includes/functions/functions_lookups.php [hunk 1]
```diff
@@ -336,7 +336,7 @@ function zen_requires_attribute_selection($products_id) {
               left join " . TABLE_PRODUCTS_OPTIONS . " po on pa.options_id = po.products_options_id
               where pa.products_id = " . (int)$products_id . "
               and po.language_id = " . (int)$_SESSION['languages_id'] . "
-              group by pa.options_id, pa.products_attributes_id";
+              group by pa.options_id";
```
Comment (drbyte) [inline includes/functions/functions_lookups.php]: Just checking: Which MySQL version did you test this against? 

I added the extra `group by` field as part of #662 via https://github.com/scottcwilson/zc-v1-series/commit/1399f1ebcaea4f3f5dfb874c45a86be59f08e897

Comment (scottcwilson) [inline includes/functions/functions_lookups.php]: Version 5.5.42.

Comment (drbyte) [inline includes/functions/functions_lookups.php]: Okay. I need to understand how you arrived at the conclusion that this `group by` was causing a problem.

... cuz with MySQL 5.7 (just tested on 5.7.13) without it an error occurs: 

> Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column `pa.products_attributes_id` which is not functionally dependent on columns in `GROUP BY `clause; this is incompatible with sql_mode=`only_full_group_by`

Comment (drbyte) [inline includes/functions/functions_lookups.php]: Also: the comments in this commit may help if you want to add an undocumented parameter to your `configure.php` to force a certain SQL mode for testing:
https://github.com/zencart/zencart/commit/f456c766bb69575dce47b2776463400b2c42f4ac

**Comment by scottcwilson** [comment]
When you had the extra group by, how did you test?   Did it prohibit the add of the two existing items in the test data (downloads with attributes which have 2 possible values)?  In my test this was permitted and did not work correctly.  But I can try some other modes and see if I can get a solution that works for all modes.

**Comment by drbyte** [comment]
I simply ran the query directly on the database:

```sql
select pa.products_attributes_id, pa.options_id, count(pa.options_values_id) as number_of_choices, po.products_options_type as options_type
               from products_attributes pa
                left join products_options po on pa.options_id = po.products_options_id
                where pa.products_id = 100
                and po.language_id = 1
              group by pa.options_id, pa.products_attributes_id;
```

Didn't get as far as testing the feature in ZC: if the query gives an error then PHP won't go further.

**Comment by scottcwilson** [comment]
I didn't mean my change; I meant the change you did when you added the extra group by field as part of #662 via 1399f1e.  That didn't work at all in my testing in Zen Cart; any product could be added to cart even when multiple values were possible for the product's attributes. 

Now I'm thinking maybe abandon group by altogether and just use a much simpler algorithm: 

- retrieve all possible attribute id's for the product in question
- ensure that each ID has only one possible value. 

This should be immune from version dependency.

File: zc_install/sql/demo/mysql_demo.sql [hunk 2]
```diff
@@ -398,7 +398,8 @@ INSERT INTO products (products_id, products_type, products_quantity, products_mo
 (176, 1, '1000', 'Normal', 'small_00.jpg', '100.0000', 0, '2004-10-05 16:45:25', '2004-10-05 16:47:22', NULL, '2.00', 1, 1, 0, '0', '1', '1', 0, 0, 0, 1, 0, 1, '0', 0, 1, 0, '100.0000', 55, 1),
 (177, 1, '1000', 'Special', '2_small.jpg', '100.0000', 0, '2004-10-05 16:47:45', '2004-10-06 00:05:48', NULL, '2.00', 1, 1, 0, '0', '1', '1', 0, 0, 0, 1, 0, 1, '0', 0, 1, 1, '75.0000', 55, 1),
 (179, 1, '1000', 'DOWNLOAD1', '1_small.jpg', '39.0000', 0, '2004-10-06 00:08:33', '2004-10-06 00:18:51', NULL, '0.00', 1, 1, 0, '0', '1', '1', 0, 0, 0, 1, 0, 1, '0', 0, 0, 0, '39.0000', 60, 1),
-(178, 1, '1000', 'Normal', '1_small.jpg', '60.0000', 0, '2004-10-05 16:54:52', '2004-10-05 17:15:02', NULL, '2.00', 1, 1, 0, '0', '1', '1', 0, 0, 0, 0, 0, 1, '0', 0, 1, 0, '50.0000', 55, 1);
+(178, 1, '1000', 'Normal', '1_small.jpg', '60.0000', 0, '2004-10-05 16:54:52', '2004-10-05 17:15:02', NULL, '2.00', 1, 1, 0, '0', '1', '1', 0, 0, 0, 0, 0, 1, '0', 0, 1, 0, '50.0000', 55, 1), 
+(180, 4, '1000', 'DPT', 'samples/4_small.jpg', '0.9346', 0, '2004-07-12 15:32:40', '2004-07-12 17:46:49', NULL, '0.00', 1, 1, 0, '0', '1', '1', 0, 0, 0, 1, 0, 1, '0', 40, 0, 0, '0.9300', 64, 1);
```
Comment (drbyte) [inline zc_install/sql/demo/mysql_demo.sql]: Small error. The `64` here should be `63`, as I believe you intended its master_categories_id to match the products_to_categories value you set later.

**Comment by drbyte** [comment]
> - retrieve all possible attribute id's for the product in question
- ensure that each ID has only one possible value.

That's what the `zen_requires_attribute_selection()` function is doing. The group-by simply takes care of reducing the returned data so the PHP has less it has to sort through.

**Comment by drbyte** [comment]
Simplifying the query is possible though. 

Let's try this instead:

```sql
select count(pa.options_values_id) as number_of_choices, po.products_options_type as options_type
               from products_attributes pa
                left join products_options po on pa.options_id = po.products_options_id
                where pa.products_id = 134
                and po.language_id = 1
group by options_type
```

ie:
```php
    $query = "select count(pa.options_values_id) as number_of_choices, po.products_options_type as options_type
              from " . TABLE_PRODUCTS_ATTRIBUTES . " pa
              left join " . TABLE_PRODUCTS_OPTIONS . " po on pa.options_id = po.products_options_id
              where pa.products_id = " . (int)$products_id . "
              and po.language_id = " . (int)$_SESSION['languages_id'] . "
              group by options_type";
```

**Comment by drbyte** [comment]
Updated:
```php
    $query = "select products_options_id, count(pa.options_values_id) as number_of_choices, po.products_options_type as options_type
              from " . TABLE_PRODUCTS_ATTRIBUTES . " pa
              left join " . TABLE_PRODUCTS_OPTIONS . " po on pa.options_id = po.products_options_id
              where pa.products_id = " . (int)$products_id . "
              and po.language_id = " . (int)$_SESSION['languages_id'] . "
              group by products_options_id, options_type";
```

**Comment by drbyte** [comment]
The updated query posted above appears to resolve the group_by issue and the #1082 issue.

**Comment by scottcwilson** [comment]
Tested - works successfully to prevent adds from listing page that won't work, and correctly processes adds that will work.

**Comment by drbyte** [comment]
👍

**Comment by zcwilt** [comment]
:+1:

**Pull request merged**